client.go

```diff
@@ -122,61 +144,85 @@ func (c *Client) FetchKeyQuota(ctx context.Context, accessKey, origin string, no
+			if errors.Is(err, ErrCachePing) {
+				if _, err := c.quotaClient.PrepareUsage(ctx, quota.AccessKey.ProjectID, quota.Cycle, now); err != nil {
+					logger.Error("unexpected client error", slog.Any("err", err))
+					if _, err := c.usageCache.ClearComputeUnits(ctx, key); err != nil {
+						logger.Error("unexpected cache error", slog.Any("err", err))
+					}
+					return 0, nil
+				}
+				continue
 			}
-			if !ok {
-				return 0, proto.ErrTimeout
+
+			// wait for cache to be ready
+			if errors.Is(err, ErrCacheWait) {
+				time.Sleep(time.Millisecond * 100 * (i + 1))
+				continue
 			}
-			fallthrough
-		case ErrCacheWait:
-			time.Sleep(time.Millisecond * 100 * (i + 1))
-		default:
+
+			logger.Error("unexpected cache error", slog.Any("err", err))
 			return 0, err
 		}
+
+		return usage, nil
 	}
-	return 0, proto.ErrTimeout
+	logger.Error("operation timed out")
+	return 0, nil
 }
 
-func (c *Client) FetchUserPermission(ctx context.Context, projectID uint64, userID string, useCache bool) (*proto.UserPermission, *proto.ResourceAccess, error) {
-	var userPerm *proto.UserPermission
-	var resourceAccess *proto.ResourceAccess
-	var err error
-
+// FetchPermission fetches the user permission from cache or from the quota server.
+// If an error occurs, it returns nil.
+func (c *Client) FetchPermission(ctx context.Context, projectID uint64, userID string, useCache bool) (proto.UserPermission, *proto.ResourceAccess, error) {
+	logger := c.logger.With(
+		slog.String("op", "spend_quota"),
+		slog.Uint64("project_id", projectID),
+		slog.String("user_id", userID),
+	)
 	// Check short-lived cache if requested. Note, the cache ttl is 10 seconds.
 	if useCache {
-		userPerm, resourceAccess, err = c.permCache.GetUserPermission(ctx, projectID, userID)
+		perm, access, err := c.permCache.GetUserPermission(ctx, projectID, userID)
 		if err != nil {
 			// log the error, but don't stop
-			c.logger.With("err", err).Error("FetchUserPermission failed to query the permCache")
+			logger.Error("unexpected cache error", slog.Any("err", err))
+		}
+		if perm != proto.UserPermission_UNAUTHORIZED {
+			return perm, access, nil
 		}
 	}
 
 	// Ask quotacontrol server via client
-	if userPerm == nil {
-		userPerm, resourceAccess, err = c.quotaClient.GetUserPermission(ctx, projectID, userID)
-		if err != nil {
-			return userPerm, resourceAccess, err
-		}
+	perm, access, err := c.quotaClient.GetUserPermission(ctx, projectID, userID)
+	if err != nil {
+		logger.Error("unexpected client error", slog.Any("err", err))
+		return proto.UserPermission_UNAUTHORIZED, nil, err
 	}
+	return perm, access, nil
+}
 
-	// Check if userPerm is still nil, in which case return unauthorized
-	if userPerm == nil {
-		v := proto.UserPermission_UNAUTHORIZED
-		return &v, resourceAccess, nil
+func (c *Client) SpendQuota(ctx context.Context, quota *proto.AccessQuota, computeUnits int64, now time.Time) (bool, error) {
+	// quota is nil only on unexpected errors from quota fetch
```

Comment:
maybe log some error, if it is unexpected, ignore this, if it is already logged somewhere else.